hydromt/components/config.py

```diff
@@ -0,0 +1,215 @@
+"""A component to write configuration files for simulations/kernels."""
+
+from os import makedirs
+from os.path import abspath, dirname, isabs, isfile, join, splitext
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict, Optional
+
+from hydromt.components.base import ModelComponent
+from hydromt.hydromt_step import hydromt_step
+from hydromt.io.path import make_config_paths_relative
+from hydromt.io.readers import read_yaml
+from hydromt.io.writers import write_toml, write_yaml
+
+if TYPE_CHECKING:
+    from hydromt.models import Model
+
+DEFAULT_CONFIG_PATH = "config.yml"
+
+
+class ConfigComponent(ModelComponent):
+    """A component to write configuration files for simulations/kernels."""
+
+    def __init__(self, model: "Model", config_fn: Optional[str] = None):
+        """Initialize a ConfigComponent.
+
+        Parameters
+        ----------
+        model: Model
+            HydroMT model instance
+        """
+        self._data: Optional[Dict[str, Any]] = None
+        if config_fn is not None:
+            self._config_fn: str = config_fn
+        else:
+            self._config_fn = DEFAULT_CONFIG_PATH
+
+        super().__init__(model=model)
+
+    @property
+    def data(self) -> Dict[str, Any]:
+        """Kernel config values."""
+        if self._data is None:
+            self._initialize()
+        if self._data is None:
+            raise RuntimeError("Could not load data for kernel config component")
+        else:
+            return self._data
+
+    def _initialize(self, skip_read=False) -> None:
+        """Initialize the kernel configs."""
+        if self._data is None:
+            self._data = dict()
+            if not skip_read:
+                self.read()
+
+    @hydromt_step
+    def write(
+        self,
+        path: Optional[str] = None,
+    ) -> None:
+        """Write kernel config at <root>/{path}."""
+        self._root._assert_write_mode()
+        if self.data:
+            if path is not None:
+                p = join(self._root.path, path)
+            else:
+                p = join(self._root.path, self._config_fn)
+
+            write_path = join(self._root.path, p)
+            self._model.logger.info(f"Writing kernel config to {write_path}.")
+            makedirs(dirname(write_path), exist_ok=True)
+
+            write_data = make_config_paths_relative(self.data, self._root.path)
+            ext = splitext(p)[-1]
+            if ext in [".yml", ".yaml"]:
+                write_yaml(write_path, write_data)
+            elif ext == ".toml":
+                write_toml(write_path, write_data)
+            else:
+                raise ValueError(f"Unknown file extention: {ext}")
+
+        else:
+            self._model.logger.debug("No kernel config found, skip writing.")
+
+    @hydromt_step
+    def read(self, path: Optional[str] = None) -> None:
```

Comment:
Instead of optional, I would just put DEFAULT_PATH here, like I did in my PR. Then you can get rid of config_fn in the constructor. Unless that's the way we want to do it of course.

```suggestion
    def read(self, path: str = DEFAULT_PATH) -> None:
```

Reply:
to be discussed tomorrow

Reply:
Resolving as we discussed this during the refinement today